File: 2025/02/solution2.php
#!/usr/bin/env php
<?php

$input = trim(file_get_contents('input.txt'));

// Parse ranges
$ranges = [];
foreach (explode(',', $input) as $range_str) {
    list($start_id, $end_id) = explode('-', $range_str);
    $ranges[] = [(int)$start_id, (int)$end_id];
}

function invalid_id($id) {
    $id_str = (string)$id;
    $length = strlen($id_str);

    // Try all possible pattern lengths from 1 to length/2
    // (need at least 2 repetitions, so pattern can be at most length/2)
    for ($pattern_length = 1; $pattern_length <= $length / 2; $pattern_length++) {
        // Check if length is divisible by pattern_length
        if ($length % $pattern_length != 0) {
            continue;
        }

        $repetitions = $length / $pattern_length;
        // Need at least 2 repetitions
        if ($repetitions < 2) {
            continue;
        }

        // Extract the pattern (first pattern_length characters)
        $pattern = substr($id_str, 0, $pattern_length);

        // Check if repeating the pattern repetitions times equals the whole string
        if (str_repeat($pattern, $repetitions) == $id_str) {
            return true;
        }
    }

    return false;
}

// Find all invalid IDs in all ranges
$invalid_ids = [];

foreach ($ranges as $range) {
    list($start_id, $end_id) = $range;
    for ($id = $start_id; $id <= $end_id; $id++) {
        if (invalid_id($id)) {
            $invalid_ids[] = $id;
        }
    }
}

// Sum all invalid IDs
$sum = array_sum($invalid_ids);

echo "Total invalid IDs found: " . count($invalid_ids) . "\n";
echo "Sum of invalid IDs: " . $sum . "\n";

File: 2025/02/solution2.lua
#!/usr/bin/env lua

local input = io.open('input.txt', 'r'):read('*all'):match('^%s*(.-)%s*$')

-- Parse ranges
local ranges = {}
for range_str in input:gmatch('[^,]+') do
  local start_id, end_id = range_str:match('(%d+)-(%d+)')
  table.insert(ranges, {tonumber(start_id), tonumber(end_id)})
end

local function invalid_id(id)
  local id_str = tostring(id)
  local length = #id_str

  -- Try all possible pattern lengths from 1 to length/2
  -- (need at least 2 repetitions, so pattern can be at most length/2)
  for pattern_length = 1, math.floor(length / 2) do
    -- Check if length is divisible by pattern_length
    if length % pattern_length ~= 0 then
      goto continue
    end

    local repetitions = length / pattern_length
    -- Need at least 2 repetitions
    if repetitions < 2 then
      goto continue
    end

    -- Extract the pattern (first pattern_length characters)
    local pattern = id_str:sub(1, pattern_length)

    -- Check if repeating the pattern repetitions times equals the whole string
    local repeated = ""
    for i = 1, repetitions do
      repeated = repeated .. pattern
    end

    if repeated == id_str then
      return true
    end

    ::continue::
  end

  return false
end

-- Find all invalid IDs in all ranges
local invalid_ids = {}

for _, range in ipairs(ranges) do
  local start_id, end_id = range[1], range[2]
  for id = start_id, end_id do
    if invalid_id(id) then
      table.insert(invalid_ids, id)
    end
  end
end

-- Sum all invalid IDs
local sum = 0
for _, id in ipairs(invalid_ids) do
  sum = sum + id
end

print("Total invalid IDs found: " .. #invalid_ids)
print("Sum of invalid IDs: " .. sum)
